runtime/src/accounts_db.rs

```diff
@@ -989,11 +1001,28 @@ impl AccountsDB {
     }
 
     fn shrink_stale_slot(&self, candidates: &mut MutexGuard<Vec<Slot>>) -> usize {
-        if let Some(slot) = self.do_next_shrink_slot(candidates) {
-            self.do_shrink_stale_slot(slot)
-        } else {
-            0
+        let mut total = 0;
+        let mut count = 0;
+        let start = Instant::now();
+        let num_roots = self.accounts_index.num_roots();
+        loop {
+            if let Some(slot) = self.do_next_shrink_slot(candidates) {
+                total += self.do_shrink_stale_slot(slot);
+            } else {
+                return 0;
+            }
+            if start.elapsed().as_millis() > 100 || count > num_roots / 10 {
```

Comment:
Before this change, I think @ryoqun can attest to this as well, but one of the dangers we've observed of shrinking too fast is that if the slot is not at the shrink threshold yet (< 80%), then we pop off the slot from the `candidates` set here, but not shrinking actually happens, and then that candidate potentially isn't retried for a while.

From my understanding, the idea here is to try to scan for the first set of shrinkable slots, otherwise we're processing 1/10th of the roots, and then the `do_next_shrink_slot` will re-add the roots again once the iteration finishes. I think this works because the only time a slot is removed from `roots` is through `clean_dead_slot`, in which case `purge_slots` will also have cleaned up all the storages.

Reply:
I think @carllin is right.

Notice that original shrinking impl were intentionally designed to be less aggressive concerned about ramp-tps's bursted txes. (oh, good old days.) :)

We've now encountered the real tx load. So this aggressiveness makes sense now. Also, i think this aggressiveness can evade the shinking-not-retried-for-a-while issue.